icu4c/source/test/intltest/dtifmtts.cpp

```diff
@@ -2135,4 +2137,212 @@ void DateIntervalFormatTest::testTicket20707() {
     }
 }
 
+void DateIntervalFormatTest::getCategoryAndField(
+        const FormattedDateInterval& formatted,
+        std::vector<int32_t>& categories,
+        std::vector<int32_t>& fields,
+        IcuTestErrorCode& status) {
+    categories.clear();
+    fields.clear();
+    ConstrainedFieldPosition cfpos;
+    while (formatted.nextPosition(cfpos, status)) {
+        categories.push_back(cfpos.getCategory());
+        fields.push_back(cfpos.getField());
+    }
+}
+
+void DateIntervalFormatTest::verifyCategoryAndField(
+        const FormattedDateInterval& formatted,
+        const std::vector<int32_t>& categories,
+        const std::vector<int32_t>& fields,
+        IcuTestErrorCode& status) {
+    ConstrainedFieldPosition cfpos;
+    int32_t i = 0;
+    while (formatted.nextPosition(cfpos, status)) {
+        assertEquals("Category", cfpos.getCategory(), categories[i]);
+        assertEquals("Field", cfpos.getField(), fields[i]);
+        i++;
+    }
+}
+
+void DateIntervalFormatTest::testTicket21222GregorianEraDiff() {
+    IcuTestErrorCode status(*this, "testTicket21222GregorianEraDiff");
+
+    LocalPointer<Calendar> cal(Calendar::createInstance(*TimeZone::getGMT(), status));
+    if (U_FAILURE(status)) {
+        errln("Failure encountered: %s", u_errorName(status));
+        return;
+    }
+    std::vector<int32_t> expectedCategory;
+    std::vector<int32_t> expectedField;
+
+    // Test Gregorian calendar
+    LocalPointer<DateIntervalFormat> g(
+        DateIntervalFormat::createInstance(
+            u"h", Locale("en"), status));
+    if (U_FAILURE(status)) {
+        errln("Failure encountered: %s", u_errorName(status));
+        return;
+    }
+    g->setTimeZone(*(TimeZone::getGMT()));
+    cal->setTime(Calendar::getNow(), status);
+    cal->set(123, UCAL_APRIL, 5, 6, 0);
+    FormattedDateInterval formatted;
+
+    UDate date0123Apr5AD = cal->getTime(status);
+
+    cal->set(UCAL_YEAR, 124);
+    UDate date0124Apr5AD = cal->getTime(status);
+
+    cal->set(UCAL_ERA, 0);
+    UDate date0124Apr5BC = cal->getTime(status);
+
+    cal->set(UCAL_YEAR, 123);
+    UDate date0123Apr5BC = cal->getTime(status);
+
+    DateInterval bothAD(date0123Apr5AD, date0124Apr5AD);
+    DateInterval bothBC(date0124Apr5BC, date0123Apr5BC);
+    DateInterval BCtoAD(date0123Apr5BC, date0124Apr5AD);
+
+    formatted = g->formatToValue(bothAD, status);
+    assertEquals("Gregorian - calendar both dates in AD",
+                 u"4/5/123, 6 AM \u2013 4/5/124, 6 AM",
+                 formatted.toString(status));
+
+    formatted = g->formatToValue(bothBC, status);
+    assertEquals("Gregorian - calendar both dates in BC",
+                 u"4/5/124, 6 AM \u2013 4/5/123, 6 AM",
+                 formatted.toString(status));
+
+    formatted = g->formatToValue(BCtoAD, status);
+    assertEquals("Gregorian - BC to AD",
+                 u"4 5, 123 BC, 6 AM \u2013 4 5, 124 AD, 6 AM",
```

Comment:
How come the format changed here?  I would have expected "4/5/123 BC", for example.  Do we not have that pattern?

Reply:
>  I would have expected "4/5/123 BC", for example. Do we not have that pattern?

we do NOT have
 GyMd{"M/d/y G"}
but we have
 GyMMMd{"MMM d, y G"}

Here is what I understand why we got this, 
for bothAD and bothBC, we need to display yMdBh while the largest different unit is y between the two time. notice the format is "h" 
since there are no pattern for yMdh in intervalFormats
It has to use fallback{"{0} – {1}"} inside intervalFormats to glue the two different date separated formatted by SimpleDateFormat
so now SimpleDateFormat try to format each of the time for yMdh
but there are no yMdh in availableFormats
so it has to use the the glue pattern in DateTimePatterns ("{1}, {0}") to glue the formatted value of yMd and the formatted value of Bh.
now, we found pattern for yMd as yMd{"M/d/y"} so the formatted result is "4/5/124" and "4/5/123"
and we found pattern for h  as h{"h a"} so the formatted result is "6 AM"
so the DateTimePatterns ("{1}, {0}") to glue the start time to "4/5/124, 6AM" and the end time to "4/5/123, 6AM" 
and fallback{"{0} – {1}"} glue them as "4/5/124, 6AM – 4/5/123, 6AM" 
(The fact that we do not display "BC" is another separated issue)

now, for the case of BCtoAD
we need to display GyMdBh while the largest different unit is G between the two time. notice the format is "h" 
since there are no pattern for GyMdBh in intervalFormats
It has to use fallback{"{0} – {1}"} inside intervalFormats to glue the two different date separated formatted by SimpleDateFormat (this is the same as the case of bothBC and both AD)

so now SimpleDateFormat try to format each of the time for GyMdh
but there are no GyMdh in availableFormats
so it has to use the the glue pattern in DateTimePatterns ("{1}, {0}") to glue the formatted value of GyMd and the formatted value of h.
now, we found pattern for GyMd from  GyMMMd{"MMM d, y G"}  so the formatted result is "4 5, 123 BC" and "4 5, 124 AD".   [hum.... we should get "April 5, 123 BC" and "April 5, 124 AD" in that case I think....)
and we found  pattern for h  as h{"h a"} so the formatted result is "6 AM"
so the DateTimePatterns ("{1}, {0}") to glue the start time to "4 5, 123 BC, 6AM" and the end time to "4 5, 124 AD, 6AM" 
and fallback{"{0} – {1}"} glue them as "4 5, 123 BC, 6AM – 4 5, 124 AD, 6AM"

Reply:
I think the thing that's throwing me here is that the month names in the `GyMd` pattern (actually, I guess it's the `GyMMMd` pattern, based on what you're saying) are coming out as numerals instead of abbreviated month names.  If this is English with the Gregorian calendar, "4 5, 123 BC" is an extremely unnatural way to render the data.  I'm assuming we get that because we're matching the `GyMMMd` pattern and then algorithmically changing `MMM` to `M` to match what was originally requested?  But if the original pattern passed in was just `h`, who's supplying the extra fields and deciding what lengths they should have?  Something seems off here.

I think _your_ code is working correctly, so I have no problem approving this PR, but I'm wondering if it's worth filing a ticket to make the output here more consistent.